### ISSUE TITLE
Alternative that doesn't require an "agent" to be embedded in the application

### ISSUE THREAD

**Issue by cosminstirbu** [Redth/Maui.UITesting#15]
Hello,

If my understanding is correct, in order to be able to test the application you'll need to include the agent in the application code itself (e.g., `<ProjectReference Include="..\..\src\Agent\Agent.csproj" />` from [here](https://github.com/Redth/Maui.UITesting/blob/b602c1d847124adcfd17c4146093a6db758d5d5a/samples/SampleMauiApp/SampleMauiApp.csproj#L43))

I'd say this is a significant drawback, since ideally you'd want to test exactly the same binary that you also manually smoked tested and that you'll ship on the App Store/Play Store, and you wouldn't want to include the agent in the store binary. Testing the same binary gives you increased confidence, as it reduces the risk introduced by the differences between the "test" binary and the store binary.

This is why Appium has been our tool of choice for a long time, because you test the same binary that you ship on the store.

Regards,
Cosmin

**Comment by dgandhi17** [comment]
what framework do you use for your apps?

**Comment by cdefgah** [comment]
@cosminstirbu Could you please clarify, is it possible to use Appium for C#/MAUI native mobile apps? Or it works only for mobile web apps? Thank you.

**Comment by cosminstirbu** [comment]
Appium can be used for both native apps (native iOS / Android and even those built with MAUI or React Native) and hybrid (native + web).

**Comment by cdefgah** [comment]
@cosminstirbu thank you :)

**Comment by Redth** [comment]
Yes, this was something I have thought quite a bit about and ultimately decided that it's worth including the agent.

Depending on your requirements in testing, it may be next to impossible to test the 'exact' binary you want to ship.  Ideally yes, this would be the way, but the drawbacks outweigh the reality here in my opinion.

Without an agent:
- No way to call backdoor methods on many of the drivers (except android espresso basically)
- No way to query the MAUI view hierarchy directly (always dealing with platform specific differences)
- Android uses up properties that should be reserved for screen readers to create accessible apps for automation identifiers (technically you can set a custom tag and use a [view matcher](https://github.com/appium/appium-dotnet-driver/pull/447) in espresso to find by it instead)
- Android's espresso driver has proved very challenging to use reliably
  - Requires a side-loaded server apk
  - Easy to run into [gradle errors](https://github.com/appium/appium-espresso-driver/issues/639#issuecomment-757015647) with different versions of androids dependencies
  - Required to be signed with the same key as the side loaded server (so either you build the server yourself or re-sign your binary which kinda defeats the purpose of your request to test the exact binary you'd release)

At the end of the day, it's much more practical to require an agent, and while I agree it would be great to have the binary you're testing not include it, the risk is worth the reward.  Xamarin.UITest always required calabash on iOS for instance, and that didn't seem to be a deal breaker for most.  

There's ways to mitigate this risk.  For instance, maybe have CI build both versions of the .apk to minimize the chance that some build environment difference is a factor in producing a potentially broken test apk vs the one you want to ship.  Of course this doesn't mitigate the risk completely, but it should be pretty close.

My decision is ultimately to go this route for the added features and reliability.  I personally have no plans to do this any differently, however it's certainly possible to create a driver implementation for this library in a manner you're after, even on top of Appium if you really wanted.

Hope that helps clear up why this is the path I've chosen :)

**Comment by cdefgah** [comment]
@Redth thank you for clarification.

**Comment by cosminstirbu** [comment]
@Redth - thank you for the clarification.

Assuming this experiment will become part of the official product, it will be useful to highlight the implications of this design choice in the documentation so developers are aware of the differences between the tested binary and the store binary.

For instance, they might be tempted to not enable linking/trimming in the tested binary, while it will be enabled on the store binary. This can easily lead to shipping linking/trimming bugs on the store.
